### PR TITLE
feat: add rate limit in log and StatusCodeError

### PR DESCRIPTION
**Pull request by dominique-pfister**
related to: https://github.com/adobe/helix-admin/issues/1267

### CODE REVIEW

File: src/RateLimit.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    return this._retryAfter;
+  }
+
+  toString() {
+    let s = `\n  limit: ${this.limit}\n  remaining: ${this.remaining}\n  reset: ${this.reset}`;
+    if (this.retryAfter) {
+      s = `${s}\n  retry after: ${this.retryAfter}`;
+    }
+    return s;
+  }
```
Comment (dominique-pfister) [inline src/RateLimit.js:72]: conventient, still not sure whether to include LFs in the string representation

File: src/excel/NamedItemContainer.js [hunk 2]
```diff
@@ -50,7 +50,7 @@ export class NamedItemContainer {
       });
     } catch (e) {
       if ((e.details && e.details.code === 'ItemAlreadyExists') && e.statusCode !== 409) {
-        throw new StatusCodeError(e.message, 409, e.code, e);
+        throw new StatusCodeError(e.message, 409);
```
Comment (dominique-pfister) [inline src/excel/NamedItemContainer.js:53]: this was actually badly constructed (there was never a 4th arg to the `StatusCodeError` constructor), but still the test succeeded

**Review by tripodsan — APPROVED** [review]

File: src/RateLimit.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  get reset() {
+    return this._reset;
+  }
+
+  get retryAfter() {
+    return this._retryAfter;
+  }
+
+  toString() {
+    let s = `\n  limit: ${this.limit}\n  remaining: ${this.remaining}\n  reset: ${this.reset} seconds`;
```
Comment (tripodsan) [inline src/RateLimit.js:67]: wouldn't this better be a json like string?

**Review by tripodsan — CHANGES_REQUESTED** [review]
we should also return the 429 and the `retry-after` from the admin response

**Comment by tripodsan** [comment]
> we should also return the 429 and the `retry-after` with the admin response

never mind... I was confused :-)

File: src/RateLimit.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    const o = {
+      limit: this.limit,
+      remaining: this.remaining,
+      reset: this.reset,
+    };
+    if (this.retryAfter) {
+      o.retryAfter = this.retryAfter;
+    }
+    return JSON.stringify(o);
+  }
```
Comment (tripodsan) [inline src/RateLimit.js:76]:
```suggestion
  toJSON() {
    const o = {
      limit: this.limit,
      remaining: this.remaining,
      reset: this.reset,
    };
    if (this.retryAfter) {
      o.retryAfter = this.retryAfter;
    }
    return o;
  }
  
  toString() {
      return JSON.stringify(this);
  }
```

File: src/OneDrive.js [hunk 5]
```diff
@@ -128,12 +129,18 @@ export class OneDrive {
     try {
       const { fetch } = this.fetchContext;
       const resp = await fetch(url, opts);
+      const rateLimit = RateLimit.fromHeaders(resp.headers);
+
+      if (rateLimit) {
+        this.log.warn(`Rate limit reported: ${rateLimit.toString()} for tenant: ${this.auth.tenant}`);
```
Comment (tripodsan) [inline src/OneDrive.js:135]:
```suggestion
        this.log.warn({ sharepointRateLimit: { tenant: this.auth.tenant, ...rateLimit.toJSON() }});
```

File: src/RateLimit.js (hunk 4 above)
Comment (tripodsan) [reply src/RateLimit.js:76]: the `toString()` is probably not needed anymore

**Pull request merged**